src/main/scala/com/github/mrpowers/spark/fast/tests/DataframeUtil.scala

```diff
@@ -0,0 +1,115 @@
+package com.github.mrpowers.spark.fast.tests
+
+import com.github.mrpowers.spark.fast.tests.ufansi.Color.{DarkGray, Green, Red}
+import com.github.mrpowers.spark.fast.tests.ufansi.Str
+import org.apache.commons.lang3.StringUtils
+import org.apache.spark.sql.Row
+
+object DataframeUtil {
+
+  def showDataframeDiff(
+      header: (String, String),
+      actual: Array[Row],
+      expected: Array[Row],
+      truncate: Int = 20
+  ): String = {
+
+    val sb = new StringBuilder
+
+    val a = actual.zipAll(expected, Row(), Row())
+    val diff = a.map { case (a, e) =>
+      if (equals(a, e)) {
+        List(ufansi.Color.DarkGray(a.toString()), ufansi.Color.DarkGray(e.toString()))
+      } else {
+        val actual   = a.toSeq
+        val expected = e.toSeq
+        if (actual.isEmpty)
+          List(
+            Red("[]"),
+            Green(expected.mkString("[", ",", "]"))
+          )
+        else if (expected.isEmpty)
+          List(Red(actual.mkString("[", ",", "]")), Green("[]"))
+        else {
+          val withEquals = actual
+            .zip(expected)
+            .map { case (a1, e1) => (a1, e1, a1 == e1) }
+          val allFieldsAreNotEqual = !withEquals.exists(_._3)
+          if (allFieldsAreNotEqual) {
+            List(
+              Red(actual.mkString("[", ",", "]")),
+              Green(expected.mkString("[", ",", "]"))
+            )
+          } else {
+            val d = withEquals
+              .map { case (a1, e1, equal) =>
+                if (equal)
+                  (DarkGray(a1.toString()), DarkGray(e1.toString))
+                else (Red(a1.toString()), Green(e1.toString))
+              }
+            List(
+              DarkGray("[") ++ d.map(_._1).reduce(_ ++ DarkGray(",") ++ _) ++ DarkGray("]"),
```

Comment:
How about using foldLeft:
```scala
val (d1, d2) = withEquals
              .map { case (a1, e1, equal) =>
                if (equal)
                  (DarkGray(a1.toString), DarkGray(e1.toString))
                else (Red(a1.toString()), Green(e1.toString))
              }
              .foldLeft((DarkGray(""), DarkGray(""))) { 
                case ((leftAcc, rightAcc), (left, right)) =>
                (leftAcc ++ DarkGray(",") ++ left, rightAcc ++ DarkGray(",") ++ right)
              }

            List(
              DarkGray("[") ++ d1++ DarkGray("]"),
              DarkGray("[") ++ d2++ DarkGray("]")
            )
```

Reply:
Might be more performance too as we only iterate the result once

Reply:
Fold left produces this result
![image](https://github.com/user-attachments/assets/4b413be7-a52d-488d-b99c-ba54f5a5880b)

Reply:
ah right we don't need to add a "," on first fold. then it might has to be
```scala
val (d1, d2) = withEquals
              .map { case (a1, e1, equal) =>
                if (equal)
                  (DarkGray(a1.toString), DarkGray(e1.toString))
                else (Red(a1.toString()), Green(e1.toString))
              }
              .foldLeft((DarkGray("["), DarkGray("["))) { 
                case ((leftAcc, rightAcc), (left, right)) =>
                val sep = if (leftAcc == DarkGray("[")) DarkGray("") else DarkGray(",")
                (leftAcc ++ sep ++ left, rightAcc ++ sep ++ right)
              }

            List(
              d1 ++ DarkGray("]"),
              d2 ++ DarkGray("]")
            )
```

Reply:
@zeotuan Could you elaborate on benefits of this approach when comparing it to using reduce? Are you saying that reduce might iterate a collection multiple times? Maybe we can create a mkString method. That would probably simplify this.

Reply:
Yes, the main advantage would be to accumulate the result in a single pass.

Reply:
Hmm. I've run a [simple benchmark test](https://github.com/happytomatoe/spark-fast-tests/blob/benchmark/src/test/scala/com/github/mrpowers/spark/fast/tests/MkStrBenchmark.scala) and don't see any substantial difference. Disclaimer - I am not an expert in microbenchmarking. Probably most of the time is spent on copying strings arrays.
@zeotuan If you insist I will change the implemention to use fold left. IMO It looks more complicated. Please let me know which implementation we should use. 

Benchmark results:
```
::Benchmark mkStringFoldLeft::
cores: 8
hostname: fedora
name: OpenJDK 64-Bit Server VM
osArch: amd64
osName: Linux
vendor: Eclipse Adoptium
version: 17.0.12+7
Parameters(size -> 100): 0.947135 ms
Parameters(size -> 200): 3.770976 ms
Parameters(size -> 300): 8.364387 ms
Parameters(size -> 400): 14.998031 ms
Parameters(size -> 500): 24.261618 ms
Parameters(size -> 600): 36.095199 ms
Parameters(size -> 700): 50.474673 ms
Parameters(size -> 800): 67.378861 ms
Parameters(size -> 900): 87.77736 ms
Parameters(size -> 1000): 110.155542 ms

::Benchmark mkStringReduce::
cores: 8
hostname: fedora
name: OpenJDK 64-Bit Server VM
osArch: amd64
osName: Linux
vendor: Eclipse Adoptium
version: 17.0.12+7
Parameters(size -> 100): 0.969242 ms
Parameters(size -> 200): 3.763655 ms
Parameters(size -> 300): 8.389291 ms
Parameters(size -> 400): 15.065835 ms
Parameters(size -> 500): 24.222374 ms
Parameters(size -> 600): 36.162865 ms
Parameters(size -> 700): 50.476543 ms
Parameters(size -> 800): 67.567095 ms
Parameters(size -> 900): 87.429674 ms
Parameters(size -> 1000): 110.113307 ms

```

Reply:
I've changed implementation to use foldLeft. @zeotuan Is there anything left to discuss? If no could you please approve?

Reply:
Thanks for the benchmark. I was going to check if all the comparision will cause issue when i have a chance. Let's keep the mkstring version as it's more readable

Reply:
Let's use mkString considering that it's much clearer. 
whichever approach will result in sub millisecond runtime so it might not be a big problem considering that it's used for small dataframe. 
Also here is the result from running benchmark: 100 rows 10 column on just dataframUtils  (ran on local computer so it's not the most reliable one)

[info] # JMH version: 1.32
[info] # VM version: JDK 11.0.19, OpenJDK 64-Bit Server VM, 11.0.19+7
[info] # VM invoker: C:\Program Files\Eclipse Adoptium\jdk-11.0.19.7-hotspot\bin\java.exe
[info] # VM options: -Xms512M -Xmx2048M -XX:+CMSClassUnloadingEnabled -Duser.timezone=GMT
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 10 iterations, single-shot each
[info] # Measurement: 10 iterations, single-shot each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread
[info] # Benchmark mode: Single shot invocation time
[info] # Benchmark: com.github.mrpowers.spark.fast.tests.MyBenchmark.singlePassFoldLeft
[info] # Run progress: 0.00% complete, ETA 00:00:00
[info] # Fork: 1 of 2
[info] # Warmup Iteration   1: 595088001.000 ns/op
[info] # Warmup Iteration   2: 12438100.000 ns/op
[info] # Warmup Iteration   3: 9331600.000 ns/op
[info] # Warmup Iteration   4: 11829901.000 ns/op
[info] # Warmup Iteration   5: 6280400.000 ns/op
[info] # Warmup Iteration   6: 5065201.000 ns/op
[info] # Warmup Iteration   7: 4747401.000 ns/op
[info] # Warmup Iteration   8: 4092700.000 ns/op
[info] # Warmup Iteration   9: 4594000.000 ns/op
[info] # Warmup Iteration  10: 3265101.000 ns/op
[info] Iteration   1: 4137000.000 ns/op
[info] Iteration   2: 3755301.000 ns/op
[info] Iteration   3: 4121500.000 ns/op
[info] Iteration   4: 4537901.000 ns/op
[info] Iteration   5: 5907200.000 ns/op
[info] Iteration   6: 6216799.000 ns/op
[info] Iteration   7: 7696900.000 ns/op
[info] Iteration   8: 6998400.000 ns/op
[info] Iteration   9: 6770100.000 ns/op
[info] Iteration  10: 7092500.000 ns/op
[info] # Run progress: 25.00% complete, ETA 00:00:04
[info] # Fork: 2 of 2
[info] # Warmup Iteration   1: 527093000.000 ns/op
[info] # Warmup Iteration   2: 16459300.000 ns/op
[info] # Warmup Iteration   3: 17248900.000 ns/op
[info] # Warmup Iteration   4: 13900299.000 ns/op
[info] # Warmup Iteration   5: 7568500.000 ns/op
[info] # Warmup Iteration   6: 6670100.000 ns/op
[info] # Warmup Iteration   7: 4528401.000 ns/op
[info] # Warmup Iteration   8: 4707900.000 ns/op
[info] # Warmup Iteration   9: 6744899.000 ns/op
[info] # Warmup Iteration  10: 5586900.000 ns/op
[info] Iteration   1: 6263599.000 ns/op
[info] Iteration   2: 5230000.000 ns/op
[info] Iteration   3: 6461599.000 ns/op
[info] Iteration   4: 5783901.000 ns/op
[info] Iteration   5: 6194301.000 ns/op
[info] Iteration   6: 5929701.000 ns/op
[info] Iteration   7: 7967200.000 ns/op
[info] Iteration   8: 6348999.000 ns/op
[info] Iteration   9: 6972799.000 ns/op
[info] Iteration  10: 6577801.000 ns/op
[info] Result "com.github.mrpowers.spark.fast.tests.MyBenchmark.singlePassFoldLeft":
[info]   N = 20
[info]   mean = 6048175.050 ∩┐╜(99.9%) 1017452.673 ns/op
[info]   Histogram, ns/op:
[info]     [3000000.000, 3500000.000) = 0
[info]     [3500000.000, 4000000.000) = 1
[info]     [4000000.000, 4500000.000) = 2
[info]     [4500000.000, 5000000.000) = 1
[info]     [5000000.000, 5500000.000) = 1
[info]     [5500000.000, 6000000.000) = 3
[info]     [6000000.000, 6500000.000) = 5
[info]     [6500000.000, 7000000.000) = 4
[info]     [7000000.000, 7500000.000) = 1
[info]   Percentiles, ns/op:
[info]       p(0.0000) = 3755301.000 ns/op
[info]      p(50.0000) = 6240199.000 ns/op
[info]      p(90.0000) = 7636460.000 ns/op
[info]      p(95.0000) = 7953685.000 ns/op
[info]      p(99.0000) = 7967200.000 ns/op
[info]      p(99.9000) = 7967200.000 ns/op
[info]      p(99.9900) = 7967200.000 ns/op
[info]      p(99.9990) = 7967200.000 ns/op
[info]      p(99.9999) = 7967200.000 ns/op
[info]     p(100.0000) = 7967200.000 ns/op
[info] # JMH version: 1.32
[info] # VM version: JDK 11.0.19, OpenJDK 64-Bit Server VM, 11.0.19+7
[info] # VM invoker: C:\Program Files\Eclipse Adoptium\jdk-11.0.19.7-hotspot\bin\java.exe
[info] # VM options: -Xms512M -Xmx2048M -XX:+CMSClassUnloadingEnabled -Duser.timezone=GMT
[info] # Blackhole mode: full + dont-inline hint
[info] # Warmup: 10 iterations, single-shot each
[info] # Measurement: 10 iterations, single-shot each
[info] # Timeout: 10 min per iteration
[info] # Threads: 1 thread
[info] # Benchmark mode: Single shot invocation time
[info] # Benchmark: com.github.mrpowers.spark.fast.tests.MyBenchmark.twoPassesReduceMkString
[info] # Run progress: 50.00% complete, ETA 00:00:02
[info] # Fork: 1 of 2
[info] # Warmup Iteration   1: 555858701.000 ns/op
[info] # Warmup Iteration   2: 17234400.000 ns/op
[info] # Warmup Iteration   3: 19125500.000 ns/op
[info] # Warmup Iteration   4: 11127900.000 ns/op
[info] # Warmup Iteration   5: 10829800.000 ns/op
[info] # Warmup Iteration   6: 8334500.000 ns/op
[info] # Warmup Iteration   7: 8050701.000 ns/op
[info] # Warmup Iteration   8: 9434700.000 ns/op
[info] # Warmup Iteration   9: 9200900.000 ns/op
[info] # Warmup Iteration  10: 12625400.000 ns/op
[info] Iteration   1: 13420800.000 ns/op
[info] Iteration   2: 12813300.000 ns/op
[info] Iteration   3: 11092800.000 ns/op
[info] Iteration   4: 8455000.000 ns/op
[info] Iteration   5: 8597101.000 ns/op
[info] Iteration   6: 9272700.000 ns/op
[info] Iteration   7: 8920099.000 ns/op
[info] Iteration   8: 9424999.000 ns/op
[info] Iteration   9: 11157999.000 ns/op
[info] Iteration  10: 7927300.000 ns/op
[info] # Run progress: 75.00% complete, ETA 00:00:01
[info] # Fork: 2 of 2
[info] # Warmup Iteration   1: 615444300.000 ns/op
[info] # Warmup Iteration   2: 14591099.000 ns/op
[info] # Warmup Iteration   3: 17807401.000 ns/op
[info] # Warmup Iteration   4: 8378400.000 ns/op
[info] # Warmup Iteration   5: 8327500.000 ns/op
[info] # Warmup Iteration   6: 7173399.000 ns/op
[info] # Warmup Iteration   7: 5924601.000 ns/op
[info] # Warmup Iteration   8: 7134700.000 ns/op
[info] # Warmup Iteration   9: 8333099.000 ns/op
[info] # Warmup Iteration  10: 7846000.000 ns/op
[info] Iteration   1: 7324700.000 ns/op
[info] Iteration   2: 10935101.000 ns/op
[info] Iteration   3: 7927600.000 ns/op
[info] Iteration   4: 5883299.000 ns/op
[info] Iteration   5: 5967499.000 ns/op
[info] Iteration   6: 5802800.000 ns/op
[info] Iteration   7: 5789101.000 ns/op
[info] Iteration   8: 6063999.000 ns/op
[info] Iteration   9: 5255400.000 ns/op
[info] Iteration  10: 5136600.000 ns/op
[info] Result "com.github.mrpowers.spark.fast.tests.MyBenchmark.twoPassesReduceMkString":
[info]   N = 20
[info]   mean = 8358409.850 ∩┐╜(99.9%) 2195623.076 ns/op
[info]   Histogram, ns/op:
[info]     [ 5000000.000,  5500000.000) = 2
[info]     [ 5500000.000,  6000000.000) = 4
[info]     [ 6000000.000,  6500000.000) = 1
[info]     [ 6500000.000,  7000000.000) = 0
[info]     [ 7000000.000,  7500000.000) = 1
[info]     [ 7500000.000,  8000000.000) = 2
[info]     [ 8000000.000,  8500000.000) = 1
[info]     [ 8500000.000,  9000000.000) = 2
[info]     [ 9000000.000,  9500000.000) = 2
[info]     [ 9500000.000, 10000000.000) = 0
[info]     [10000000.000, 10500000.000) = 0
[info]     [10500000.000, 11000000.000) = 1
[info]     [11000000.000, 11500000.000) = 2
[info]     [11500000.000, 12000000.000) = 0
[info]     [12000000.000, 12500000.000) = 0
[info]     [12500000.000, 13000000.000) = 1
[info]     [13000000.000, 13500000.000) = 1
[info]   Percentiles, ns/op:
[info]       p(0.0000) = 5136600.000 ns/op
[info]      p(50.0000) = 8191300.000 ns/op
[info]      p(90.0000) = 12647769.900 ns/op
[info]      p(95.0000) = 13390425.000 ns/op
[info]      p(99.0000) = 13420800.000 ns/op
[info]      p(99.9000) = 13420800.000 ns/op
[info]      p(99.9900) = 13420800.000 ns/op
[info]      p(99.9990) = 13420800.000 ns/op
[info]      p(99.9999) = 13420800.000 ns/op
[info]     p(100.0000) = 13420800.000 ns/op
[info] # Run complete. Total time: 00:00:05
[info] REMEMBER: The numbers below are just data. To gain reusable insights, you need to follow up on
[info] why the numbers are the way they are. Use profilers (see -prof, -lprof), design factorial
[info] experiments, perform baseline and negative tests that provide experimental control, make sure
[info] the benchmarking environment is safe on JVM/OS/HW level, ask for reviews from the domain experts.
[info] Do not assume the numbers tell you what you want them to tell.
[info] Benchmark                            Mode  Cnt        Score         Error  Units
[info] MyBenchmark.singlePassFoldLeft         ss   20  6048175.050 ∩┐╜ 1017452.673  ns/op
[info] MyBenchmark.twoPassesReduceMkString    ss   20  8358409.850 ∩┐╜ 2195623.076  ns/op

Reply:
@zeotuan Done. Could you please check and approve or is there something left to discuss?

Reply:
I don't see any problem with the use of reduce if it guarantees that the list is not empty. 
If we want to check if there's an improvement in performance maybe try to avoid creating the middle lists and go for a view / iterator approach, but let's see that in other different pr